### PR TITLE
Add PhayaThaiBERT engine with new features [WIP]

### LINKED ISSUE #868: Add PhayaThaiBERT model into PyThaiNLP [WIP]
Due to an impressive result of the new released paper **PhayaThaiBERT: Enhancing a Pretrained Thai Language Model with Unassimilated Loanwords** for better handling with foreign words compared to former existing Thai encoder-based model. 

I think it is great to add it into supported downstream task of PyThaiNLP e.g. token classification etc. to strengthen the library. What do you think? If all of us agreed on this, I can help integrating it as a new engine asap.

## New features 
Here is the task that I found that it can be integrated in PyThaiNLP after reading a paper. The list below here is the current progress and contributors who put their efforts develop the model ( ✅  check mark means that it already added in the source code and will make a complete PR after complete all of it krub):

- [x] Part-of-speech tagging on blackboard corpus by @MpolaarbearM 
- [x] Named-entity-recognition on Thainer-v2 corpus by @pavaris-pm 
- [x] Tokenization by @pavaris-pm 
- [x] Data Augmentation (Text) by @pavaris-pm 
- [ ] Word Correction (currently under research and development)

etc ... (I will keep add more into the list based on what I have found during an experiment)

For those who interested, feel free to leave a comment below in case you want to develop a model in any of your interested task krub. After that, you can made a PR to the same brach as in PR #873

### PR DESCRIPTION
**Pull request by pavaris-pm**
## What does this changes

According to #868, i have made a new PR by added new folder named `phayathaibert` since it is a new Thai language model, with that, i decided to treat it to be the same as we treat `wangchanberta` because it also has their own folder as well. Apart from introducing a new model, i currently made an experiment and added new features into pythainlp including its test cases as well (you can see the list in this PR description). Clearly note here that this PR also solved the unsync forked version in #871 as well.

Will resolve #871 and fix #868.

###  ⚠️ [NOTE] i will ask for a review after all tasks in the below list completed
Here is the task which I found that PhayaThaiBERT can be integrated in PyThaiNLP after reading a paper. The list below here is the current progress (check mark means that I already added in the source code and will ask for your review after i complete all of them krub):

- [x] Part-of-speech tagging
- [x] Named-entity-recognition
- [x] Tokenization
- [x] Data Augmentation (Text)

## Upcoming features that can be added soon (futurePR)
- [ ] Word Correction  (Under heavy development)

etc ... (I will keep add more into the list based on what I have found during an experiment)

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](https://github.com/PyThaiNLP/pythainlp/blob/dev/CONTRIBUTING.md) to this repository.

- [x] Passed code styles and structures
- [ ] Passed code linting checks and unit test

### CODE REVIEW

File: pythainlp/phayathaibert/core.py
Comment (pavaris-pm) [inline pythainlp/phayathaibert/core.py:1]: here is the newly added folder for **phayathaibert** same as what we do with **wangchanberta** krub.

Comment (bact) [reply pythainlp/phayathaibert/core.py:1]: This is something I'm not sure about current convention, in terms of directory structure.

In general, it used to be the case that we

- use verbs for functions (like we have renamed `romanization()` to `romanize()` #133)
- trying to use verbs for modules that provide functions: `pythainlp.tokenize`, `pythainlp.augment`
  - in some cases, modules that provide functions are named after the name of the task: `pythainlp.coref`
- trying to use nouns for modules that provide resources or a collections of miscellaneous utility: `pythainlp.corpus`, `pythainlp.util`, `pythainlp.tools`
- trying to be generic (like we have moved `ner` to `pythainlp.tag` #160)

These are more or less to follow the convention in NTLK https://github.com/nltk/nltk/

However, in more recent versions these naming convention are not kept very strict.

We also have a growing number of modules for a specific language model (ulmfit, word2vec, wangchanberta) and also their corresponding subdirectories.

(Keep in mind that module structure doesn't always have to follow directory structure. We can have a very deep directory structure and manage to import them internally to expose (using `__all__`) them as a 2-level package.module structure).

I think the current situation is not bad, with few LMs. But if in the future we're going to support more LMs, maybe we need to think more about the structure. Where to put them nicely?

For example, should each of the LMs has their own dir (as it is now), or should we group them together and put them inside the existing `corpus`, or should the group has their own new `lm` or `langmodels` subdir?

--

In any case, from the user's point of view, we should stick to only 2-level `package.module` and not deeper (like `package.subpackage.module` since that's against general Python convention).

For example, for `pythainlp.util.normalize.remove_dup_spaces()`,
user can just do
```python
from pythainlp.util import remove_dup_spaces
```
without the need to
```python
from pythainlp.util.normalize import remove_dup_spaces
```

since `pythainlp.util.__init__` have done he re-export of `remove_dup_spaces` for us already.

Comment (bact) [reply pythainlp/phayathaibert/core.py:1]: @wannaphong and @MpolaarbearM  may have some thoughts.

Comment (pavaris-pm) [reply pythainlp/phayathaibert/core.py:1]: @bact I'm really much agree with it since it seems to be more deeper directory structure when more functions/tools were added in, which is unnecessary doing it.  With that, a discussion is much required in order to setup a new general convention from now on. After that, we will begin to re-structure to make it fit for our new convention krub.

In my opinion, to handling with LMs, I think it is great to introduce a new directory like e.g. `models`,`lm` or `langmodels` to keep every Thai LM inside (same as **Huggingface** style) so that it can be designed/experimented separately based on their specific characteristics, and write a flexible code to make it can map with all supported corpus as what we do ,and handle with longer model names with dictionary lookup as in `pos_tag_transformers` is seems to be the leanest solution from my aspect krub.

Ps. As i told you in a previous issue about naming convention with @bact in #865 , i have no free time to bring up new discussion in this shortcoming weeks. however, i did not forget about the new discussion for naming convention that should be setup. if i have free time, i ensure that i will bring that up asap krub.

Comment (pavaris-pm) [reply pythainlp/phayathaibert/core.py:1]: Here is a dictionary lookup to handle longer model names as what i mean in `pos_tag_transformers` krub

https://github.com/PyThaiNLP/pythainlp/blob/92f022b886bd6f627f158ce9cfa72135f73c8515/pythainlp/tag/pos_tag.py#L214-L236

**Comment by pavaris-pm** [comment]
@bact @wannaphong i've already add all features that i've been found from phayathaibert  into the source code and already fix pep8 format. 
### As of today, here is new features added 
1. Part-of-speech tagging model trained by @MpolaarbearM (already made co-authored commit 👍🏻 )
2. Named-entity-recognition model trained by @pavaris-pm 
3. Tokenization (even it used the same tokenizer as WangchanBERTa, clearly note here that **Vocabulary Expansion** was integrated into it) added by @pavaris-pm 
4. Data Augmentation (Text) by @pavaris-pm 

### Upcoming features that can be added soon (future PR)
1. Word Correction (i've found a nice paper for [Two-stage Thai Misspelling Correction based on Pre-trained Language Models](https://ieeexplore.ieee.org/abstract/document/10202006) and need further research on it)
https://github.com/PyThaiNLP/pythainlp/blob/75bbc077162169466b30872c035092903fc60a78/pythainlp/spell/wanchanberta_thai_grammarly.py#L4-L11

According to this, i think that adding these 4 completed features first, and the upcoming features (e.g. word correction) can be added later with the next PR because it is better to bring the state-of-the-art Thai encoder based model into production asap. With that, you can review this PR and suggest for further development of them. If you're ok with this, you can approve and merge it krub.

**Comment by pavaris-pm** [comment]
@MpolaarbearM kindly inform here that **co-authored commit** already made krub. You can check it 😄

**Comment by bact** [comment]
There were few error test suite (not related to your PR).
I have fixed them, do you mind to sync your code with the latest from dev? So the test can be run again.
thanks.

**Comment by pavaris-pm** [comment]
> There were few error test suite (not related to your PR).
> 
> I have fixed them, do you mind to sync your code with the latest from dev? So the test can be run again.
> 
> thanks.

Roger that. I'll do it krub.

**Comment by pavaris-pm** [comment]
@bact I'm done with syncing already

**Review by bact — APPROVED** [review]
